component/common/loki/client/queue_client.go

```diff
@@ -0,0 +1,486 @@
+		e.Line = e.Line[:c.maxLineSize]
+	}
+
+	batch, ok := c.batches[tenantID]
+
+	// If the batch doesn't exist yet, we create a new one with the entry
+	if !ok {
+		nb := newBatch(c.maxStreams)
+		_ = nb.addFromWAL(lbs, e, segment)
+		c.batches[tenantID] = nb
+		c.initBatchMetrics(tenantID)
+		return
+	}
+
+	// If adding the entry to the batch will increase the size over the max
+	// size allowed, we do send the current batch and then create a new one
+	if batch.sizeBytesAfter(e.Line) > c.cfg.BatchSize {
+		c.enqueue(tenantID, batch)
+
+		nb := newBatch(c.maxStreams)
+		_ = nb.addFromWAL(lbs, e, segment)
+		c.batches[tenantID] = nb
+
+		return
+	}
+
+	// The max size of the batch isn't reached, so we can add the entry
+	err := batch.addFromWAL(lbs, e, segment)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "batch add err", "tenant", tenantID, "error", err)
+		reason := ReasonGeneric
+		if err.Error() == errMaxStreamsLimitExceeded {
+			reason = ReasonStreamLimited
+		}
+		c.metrics.droppedBytes.WithLabelValues(c.cfg.URL.Host, tenantID, reason).Add(float64(len(e.Line)))
+		c.metrics.droppedEntries.WithLabelValues(c.cfg.URL.Host, tenantID, reason).Inc()
+	}
+}
+
+type queuedBatch struct {
+	TenantID string
+	Batch    *batch
+}
+
+func (c *queueClient) enqueue(tenantID string, b *batch) {
+	c.sendQueue <- queuedBatch{
+		TenantID: tenantID,
+		Batch:    b,
+	}
+}
+
+func (c *queueClient) runSendQueue() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case <-c.quit:
+			return
+
+		case qb, ok := <-c.sendQueue:
+			if !ok {
+				return
+			}
+			c.sendBatch(context.Background(), qb.TenantID, qb.Batch)
+		}
+	}
+}
+
+func (c *queueClient) runSendOldBatches() {
+	// Given the client handles multiple batches (1 per tenant) and each batch
+	// can be created at a different point in time, we look for batches whose
+	// max wait time has been reached every 10 times per BatchWait, so that the
+	// maximum delay we have sending batches is 10% of the max waiting time.
+	// We apply a cap of 10ms to the ticker, to avoid too frequent checks in
+	// case the BatchWait is very low.
+	minWaitCheckFrequency := 10 * time.Millisecond
+	maxWaitCheckFrequency := c.cfg.BatchWait / 10
+	if maxWaitCheckFrequency < minWaitCheckFrequency {
+		maxWaitCheckFrequency = minWaitCheckFrequency
+	}
+
+	maxWaitCheck := time.NewTicker(maxWaitCheckFrequency)
+
+	// pablo: maybe this should be moved out
+	defer func() {
+		maxWaitCheck.Stop()
+		c.wg.Done()
+	}()
+
+	for {
+		select {
+		case <-c.quit:
+			return
+
+		case <-maxWaitCheck.C:
+			c.batchesMtx.Lock()
+			// Send all batches whose max wait time has been reached
+			for tenantID, b := range c.batches {
+				if b.age() < c.cfg.BatchWait {
+					continue
+				}
+
+				c.enqueue(tenantID, b)
+				// deleting assuming that since the batch expired the wait time, it
+				// hasn't been written for some time
+				delete(c.batches, tenantID)
+			}
+			c.batchesMtx.Unlock()
```

Comment:
Similar question here, would it be safe to put the items to enqueue in to a slice, release the batch mutex, and then enqueue the items from the slice to avoid blocking with the batch mutex held?

Reply:
absolutely yes!